File: AnalyticsKitDebugProvider.m

#import "AnalyticsKitDebugProvider.h"

@interface AnalyticsKitDebugProvider()

// borrowing from https://github.com/dbettermann/DBAlertController/blob/master/Source/DBAlertController.swift
@property(nonatomic,strong)UIWindow *alertWindow;

@end

@implementation AnalyticsKitDebugProvider

#pragma mark -
#pragma mark Lifecycle

-(void)applicationWillEnterForeground{}

-(void)applicationDidEnterBackground{}

-(void)applicationWillTerminate{}

-(void)uncaughtException:(NSException *)exception{}

#pragma mark -
#pragma mark Event Logging

-(void)logScreen:(NSString *)screenName{}

-(void)logEvent:(NSString *)value {}

-(void)logEvent:(NSString *)event withProperty:(NSString *)key andValue:(NSString *)value {}

-(void)logEvent:(NSString *)event withProperties:(NSDictionary *)dict {}

-(void)logEvent:(NSString *)eventName timed:(BOOL)timed{}

-(void)logEvent:(NSString *)eventName withProperties:(NSDictionary *)dict timed:(BOOL)timed{}

-(void)endTimedEvent:(NSString *)eventName withProperties:(NSDictionary *)dict{}

-(void)showDebugAlert:(NSString *)message{
        [[NSOperationQueue mainQueue] addOperationWithBlock:^{
            if (self.alertWindow != nil) {
                [[self.alertWindow rootViewController] dismissViewControllerAnimated:NO completion:nil];
                self.alertWindow.rootViewController = nil;
                self.alertWindow.hidden = true;
                self.alertWindow = nil;
            }

            UIWindow *alertWindow = [[UIWindow alloc] initWithFrame:[[UIScreen mainScreen] bounds]];
            self.alertWindow = alertWindow;
            alertWindow.rootViewController = [UIViewController new];
            alertWindow.backgroundColor = [UIColor clearColor];
            [alertWindow makeKeyAndVisible];

            UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"AnalyticsKit Received Error" message:message preferredStyle:UIAlertControllerStyleAlert];
            __weak AnalyticsKitDebugProvider *weakSelf = self;
            [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                weakSelf.alertWindow.rootViewController = nil;
                weakSelf.alertWindow.hidden = true;
                weakSelf.alertWindow = nil;
            }]];

            [[alertWindow rootViewController] presentViewController:alertController animated:YES completion:nil];
        }];
}

-(void)logError:(NSString *)name message:(NSString *)message exception:(NSException *)exception{
    NSString *detail = [NSString stringWithFormat:@"%@\n\n%@\n\n%@", name, message, exception];
    [self showDebugAlert:detail];
}

-(void)logError:(NSString *)name message:(NSString *)message error:(NSError *)error{
    NSString *detail = [NSString stringWithFormat:@"%@\n\n%@\n\n%@", name, message, error];    
    [self showDebugAlert:detail];
}

@end

File: AnalyticsKitDebugProvider.swift
import Foundation
import UIKit

public class AnalyticsKitDebugProvider: NSObject, AnalyticsKitProvider {
    public func applicationWillEnterForeground() {}
    public func applicationDidEnterBackground() {}
    public func applicationWillTerminate() {}
    public func uncaughtException(_ exception: NSException) {}
    public func logScreen(_ screenName: String) {}
    public func logScreen(_ screenName: String, withProperties properties: [String: Any]) {}
    public func endTimedEvent(_ event: String, withProperties properties: [String: Any]) {}

    fileprivate weak var alertController: UIAlertController?

    // Logging
    public func logEvent(_ event: String) { }
    public func logEvent(_ event: String, withProperty key: String, andValue value: String) { }
    public func logEvent(_ event: String, withProperties properties: [String: Any]) { }

    public func logEvent(_ event: String, timed: Bool) {
        logEvent(event)
    }

    public func logEvent(_ event: String, withProperties properties: [String: Any], timed: Bool) {
        logEvent(event, withProperties: properties)
    }
    
    public func logError(_ name: String, message: String?, properties: [String: Any]?, exception: NSException?) {
        let message = "\(name)\n\n\(message ?? "nil")\n\n\(propertiesString(for: properties) ?? "nil")\n\n\(exception?.description ?? "nil")"
        showAlert(message)
    }

    public func logError(_ name: String, message: String?, properties: [String: Any]?, error: Error?) {
        let message = "\(name)\n\n\(message ?? "nil")\n\n\(propertiesString(for: properties) ?? "nil")\n\n\(error?.localizedDescription ?? "nil")"
        showAlert(message)
    }

    private func propertiesString(for properties: [String: Any]?) -> String? {
        guard let properties = properties else { return nil }
        var stringArray = [String]()
        for (key, value) in properties {
            stringArray.append("\(key): \(value)")
        }
        return stringArray.joined(separator: "\n")
    }

    fileprivate func showAlert(_ message: String) {
        if Thread.isMainThread {
            showAlertController(message)
        } else {
            DispatchQueue.main.async {
                self.showAlertController(message)
            }
        }
    }

    fileprivate func showAlertController(_ message: String) {
        // dismiss any already visible alert
        if let alertController = self.alertController {
            alertController.dismiss(animated: false, completion: nil)
        }

        let alertController = UIAlertController(title: "AnalyticsKit Received Error", message: message, preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
        self.present(alertController)
        self.alertController = alertController
    }

    fileprivate func present(_ alertController: UIAlertController) {
        let keyWindow: UIWindow? = {
            if #available(iOS 15, *) {
                // Use connectedScenes to find the key window for iOS 15 and later
                return UIApplication.shared.connectedScenes
                    .compactMap { $0 as? UIWindowScene }
                    .flatMap { $0.windows }
                    .first { $0.isKeyWindow }
            } else {
                // Use windows for iOS 13 and 14, and keyWindow for iOS 12 and earlier
                return UIApplication.shared.windows.first { $0.isKeyWindow } ?? UIApplication.shared.keyWindow
            }
        }()
        if let rootVC = keyWindow?.rootViewController {
            presentFromController(alertController, controller: rootVC)
        }
    }

    fileprivate func presentFromController(_ alertController: UIAlertController, controller: UIViewController) {
        if let presentedController = controller.presentedViewController {
            presentFromController(alertController, controller: presentedController)
        } else if let navVC = controller as? UINavigationController, let visibleVC = navVC.visibleViewController {
            presentFromController(alertController, controller: visibleVC)
        } else if let tabVC = controller as? UITabBarController, let selectedVC = tabVC.selectedViewController {
            presentFromController(alertController, controller: selectedVC)
        } else {
            controller.present(alertController, animated: true, completion: nil)
        }
    }
}
